File: Java/Linked_List/merge_two_sorted_linked_list.java
/*
        // Definition for Singly-Linked List
        public class ListNode
	{

                int val;
                ListNode next;

                public ListNode()
                {

                }

                public ListNode( int val )
                {
                        this.val = val;
                }

                public ListNode( int val, ListNode next )
                {
                        this.val = val;
                        this.next = next;
                }

        }
*/

class Solution
{
	public ListNode mergeTwoLists( ListNode l1, ListNode l2 )
	{

        	ListNode head = null;
        	ListNode tail = null;

		// LOOP THROUGH LIST 1 AND ADD NODES
		while( l1 != null )
		{
			ListNode node = new ListNode( l1.val );
			if( head == null )
			{
			    head = tail = node;
			}
			else
			{
			    tail.next = node;
			    tail = node;
			}
			l1 = l1.next;
		}

		// LOOP THROUGH LIST 2 AND ADD NODES
		while( l2 != null )
		{
			ListNode node = new ListNode( l2.val );
			if( head == null )
			{
			    head = tail = node;
			}
			else
			{
			    tail.next = node;
			    tail = node;
			}
			l2 = l2.next;
		}

		// BUBBLE SORT
		ListNode current = head;
		ListNode next = null;
		int temp = 0;

		while( current != null )
		{
			next = current.next;
			while( next != null )
			{
				if( current.val > next.val )
				{
					temp = current.val;
					current.val = next.val;
					next.val = temp;
				}
				next = next.next;
			}
			current = current.next;
		}

		return head;

	}
}

File: Programming_Skills/Linked_List/Merge_Two_Sorted_Linked_List.js
// Leetcode: https://leetcode.com/problems/merge-two-sorted-lists/

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */

/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */

const mergeTwoLists = (list1, list2) => {
    if (!list1) return list2;
    if (!list2) return list1;

    // Variables Definition
    let head = null;
    let tail = null;
    let node = null;

    const parseList1 = () => {
        node = new ListNode(list1.val);
        list1 = list1.next;
    };

    const parseList2 = () => {
        node = new ListNode(list2.val);
        list2 = list2.next;
    };

    const addToLinkedList = () => {
        if (head == null) {
            head = tail = node;
        } else {
            tail.next = node;
            tail = node;
        }
    };

    // Loop until any of the list becomes null
    while (list1 && list2) (list1.val < list2.val ? parseList1() : parseList2(), addToLinkedList());

    // Add all the nodes of list1, if remaining
    while (list1) (parseList1(), addToLinkedList());

    // Add all the nodes of list2, if remaining
    while (list2) (parseList2(), addToLinkedList());

    return head;
};
